python/cudf/cudf/core/multiindex.py

```diff
@@ -838,6 +839,112 @@ def from_product(cls, arrays, names=None):
         result = cls.from_pandas(pdi)
         return result
 
+    def _poplevels(self, level):
+        """
+        Pop the specified levels from self
+        """
+        if not pd.api.types.is_list_like(level):
+            level = (level,)
+
+        ilevels = [self._level_index_from_level(lev) for lev in level]
+
+        popped_data = OrderedDict({})
+        popped_names = []
+        names = list(self.names)
+
+        for i in reversed(sorted(ilevels)):
+            names.pop(i)
+            n = self._data.names[i]
+            popped_data[n] = self._data.pop(n)
+            popped_names.insert(0, self.names[i])
+            popped_data.move_to_end(n, last=False)
```

Comment:
Since we're popping elements off as we're iterating over indices, we need to go backwards.

The `insert` is likely expensive. But `move_to_end` should be efficient whether we are moving to the beginning (`last=False`) or the end (`last=True`) of the OrderedDict.

Perhaps we could reverse both the list and the dict once, after building them?

Reply:
Connected offline, going to replace this with two loops over `ilevels` in separate directions. First one to pop from `names` in the reverse sorted direction, second one in the sorted direction to build the `popped_data` and `popped_names` objects